diff -Naur linux.old/mm/mmap.c linux.new/mm/mmap.c
--- linux.old/mm/mmap.c	2019-04-10 14:59:24.095666665 +0800
+++ linux.new/mm/mmap.c	2019-04-10 15:00:04.451667681 +0800
@@ -2178,7 +2178,7 @@
 int expand_downwards(struct vm_area_struct *vma,
 				   unsigned long address)
 {
-	int error;
+	int error = 0;
 
 	/*
 	 * We must make sure the anon_vma is allocated
@@ -2188,9 +2188,8 @@
 		return -ENOMEM;
 
 	address &= PAGE_MASK;
-	error = security_mmap_addr(address);
-	if (error)
-		return error;
+	if (address < mmap_min_addr)
+		return -EPERM;
 
 	vma_lock_anon_vma(vma);
 
